src/test/scala/cognite/spark/v1/SequenceRowsRelationTest.scala

```diff
@@ -0,0 +1,243 @@
+      _.length != 100
+    )
+    rowNumberOnly(0).get(0) shouldBe 1L
+
+    val differentOrderProjection = retryWhile[Array[Row]](
+      spark.sql(s"select num1, num2, rowNumber, str1 from sequenceRows_a order by rowNumber").collect,
+      _.length != 100
+    )
+    differentOrderProjection(0).get(2) shouldBe 1L
+    differentOrderProjection(0).get(3) shouldBe "abc"
+    differentOrderProjection(0).get(1) shouldBe 1.1
+    differentOrderProjection(0).get(0) shouldBe 6L
+
+    val oneColumn = retryWhile[Array[Row]](
+      spark.sql(s"select num2 from sequenceRows_a order by rowNumber").collect,
+      _.length != 100
+    )
+    oneColumn.map(_.get(0)) shouldBe Seq.fill(100)(1.1)
+  }
+  it should "insert NULL values" in withSequences(Seq(sequenceA)) { case Seq(sequenceId) =>
+    // num1, str1, num2
+    insertRows(
+      sequenceId,
+      spark
+        .sql("select 1 as rowNumber, 1 as num1"))
+
+    insertRows(
+      sequenceId,
+      spark
+        .sql("select 2 as rowNumber, NULL as num1, 'abc' as str1, 1 as num2"))
+
+    insertRows(
+      sequenceId,
+      spark
+        .sql("select 3 as rowNumber, 2 as num1, NULL as str1, NULL as num2"))
+
+    val rows = retryWhile[Array[Row]](
+      spark.sql(s"select * from sequenceRows_a order by rowNumber").collect,
+      _.length != 3
+    )
+    rows.map(_.getAs[Long]("rowNumber")) shouldBe Array(1, 2, 3)
+    rows.map(_.getAs[Any]("num1")) shouldBe Array[Any](1L, null, 2L)
+    rows.map(_.getAs[String]("str1")) shouldBe Array(null, "abc", null)
+    rows.map(_.getAs[Any]("num2")) shouldBe Array[Any](null, 1.0, null)
+  }
+
+  it should "insert and update rows" in withSequences(Seq(sequenceA)) { case Seq(sequenceId) =>
+    // num1, str1, num2
+    insertRows(
+      sequenceId,
+      spark
+        .sql("select 1 as rowNumber, 1 as num1, 1 as num2, 'a' as str1"))
+
+    insertRows(
+      sequenceId,
+      spark
+        .sql("select 1 as rowNumber, 2 as num1"))
+
+    val rows = retryWhile[Array[Row]](
+      spark.sql(s"select * from sequenceRows_a order by rowNumber").collect,
+      rows => rows.length != 1 || rows(0).getAs[Long]("num1") != 2
+    )
+    println(rows(0))
+  }
+
+  it should "create and read many rows" in withSequences(Seq(sequenceB)) { case Seq(sequenceId) =>
+    // exceed the page size
+    val testSize = 20 * 1000
+    spark.sparkContext
+      .parallelize(1 to testSize)
+      .toDF()
+      .createOrReplaceTempView("numbers")
+    insertRows(
+      sequenceId,
+      spark
+        .sql("select value as rowNumber, value * 6 as num1 from numbers"))
+
+    createRowsRelation(sequenceId).createOrReplaceTempView("sequenceRows_b")
+
+    val allColumns = retryWhile[Array[Row]](
+      spark.sql(s"select * from sequenceRows_b order by rowNumber").collect,
+      _.length != testSize
+    )
+    allColumns(0).schema.fieldNames shouldBe Array("rowNumber", "num1")
+    allColumns.map(_.getAs[Long]("num1")) shouldBe (1 to testSize).map(_ * 6)
+  }
+
```

Comment:
Add tests covering the pushdown filters.

Reply:
This was a very good point :) Added the test and fixed like 3 bugs in there :facepalm: